daemon/graphdriver/overlay2/overlay.go

```diff
@@ -276,29 +275,48 @@ func parseOptions(options []string) (*overlayOptions, error) {
 
 func supportsOverlay() error {
 	// Access overlay filesystem so that Linux loads it (if possible).
-	mountTarget, err := ioutil.TempDir("", "supportsOverlay2")
+	mkdirSucc := true
+	lower, err := ioutil.TempDir("", "overlay2Lower")
 	if err != nil {
-		logrus.WithError(err).WithField("storage-driver", "overlay2").Error("could not create temporary directory, so assuming that 'overlay' is not supported")
-		return graphdriver.ErrNotSupported
+		mkdirSucc = false
+	} else {
+		defer os.RemoveAll(lower)
 	}
-	/* The mounting will fail--after the module has been loaded.*/
-	defer os.RemoveAll(mountTarget)
-	unix.Mount("overlay", mountTarget, "overlay", 0, "")
 
-	f, err := os.Open("/proc/filesystems")
+	upper, err := ioutil.TempDir("", "overlay2Upper")
 	if err != nil {
-		return err
+		mkdirSucc = false
+	} else {
+		defer os.RemoveAll(upper)
 	}
-	defer f.Close()
 
-	s := bufio.NewScanner(f)
-	for s.Scan() {
-		if s.Text() == "nodev\toverlay" {
-			return nil
-		}
+	work, err := ioutil.TempDir("", "overlay2Work")
+	if err != nil {
+		mkdirSucc = false
+	} else {
+		defer os.RemoveAll(work)
+	}
+
+	merged, err := ioutil.TempDir("", "overlay2Merged")
+	if err != nil {
+		mkdirSucc = false
+	} else {
+		defer os.RemoveAll(merged)
+	}
+
+	if !mkdirSucc {
+		logrus.WithError(err).WithField("storage-driver", "overlay2").Error("could not create temporary directory, so assuming that 'overlay2' is not supported")
+		return graphdriver.ErrNotSupported
 	}
-	logger.Error("'overlay' not found as a supported filesystem on this host. Please ensure kernel is new enough and has overlay support loaded.")
-	return graphdriver.ErrNotSupported
+
+	// Attempt to perform an overlay mount to determine whether the current user (and system) can support it
+	if err := unix.Mount("overlay", merged, "overlay", 0, fmt.Sprintf("lowerdir=%s,upperdir=%s,workdir=%s", lower, upper, work)); err != nil {
```

Comment:
for overlay2, we should better test multi-lowerdir, but not a huge deal for recent kernels.

So LGTM.